Add firebase migration guide (#1378)
FusionAuth/fusionauth-site: site/docs/v1/tech/migration-guide/firebase.adoc

---
layout: doc
title: Migration From Firebase
description: How to migrate your users to FusionAuth from Firebase.
---
:page-liquid:

:migration_source_name: Firebase
:migration_source_dir: firebase

> :script_supports_social_logins: false

Reviewer: Why do we need the `/api/identity-provider/link` permission if we aren't supporting social logins?

Reviewer: It looks like you support social logins: https://github.com/FusionAuth/fusionauth-import-scripts/pull/6/files#diff-492fa36046cc43dd051ae94842e3dcd3f5fb690ecb834e2d9d7a1ec5ef43954fR73

Author: yes, social logins are supported so we toggled this to true. Not really sure what it does though tbh.

Reviewer: Thanks!

It's a switch consumed by this include file: site/docs/v1/tech/migration-guide/_social-login-note.adoc

* Projects are a high level construct which groups entities such as
users and applications together. FusionAuth calls these Tenants.
* For Firebase, Roles can be implemented using a combination of https://firebase.google.com/docs/auth/admin/custom-claims[custom claims and security rules.] FusionAuth has Roles and they are defined on an Application by Application basis.
* Refresh tokens allow JWTs to be refreshed without a user logging in.
These can be migrated using the link:/docs/v1/tech/apis/users#import-refresh-tokens[Import Refresh Tokens API].
* Firebase has many other products your applications might be using, from Cloud Firestore for data, Analytics and more. FusionAuth does not have equivalents for these products, and is focused only on authentication. You can use FusionAuth in conjunction with all the other Firebase products.

[NOTE]
====
In FusionAuth, users are explicitly mapped to applications with a link:/docs/v1/tech/core-concepts/registrations[Registration].

In Firebase, users have access to all Firebase Applications in the same project by default. You can control access by using Roles and Security Rules per Application if you need.
====

==== Identifiers 

> include::docs/v1/tech/migration-guide/_identifiers.adoc[]

Author: they moved to UUIDs in 2016, but now away again, so we tweaked this a bit. This would be a major pain for people migrating from Firebase if they need to update all external dependencies to link to the new FusionAuth UUID? Or is there a workaround we should mention here.

> include::docs/v1/tech/migration-guide/_identifiers.adoc[]

Reviewer: The typical workaround is:

* let FusionAuth own the Ids
* add a `user.data.originalId` field, with the value of the firebase id.

Everything under `user.data` is indexed and available via search and so the system can look up a user by either their FusionAuth id or the firebase Id.

Would be great to mention this workaround. You could add it below the included generic file, if it reads okay.

      "email": "user2@test.com",
      "emailVerified": false,
      "passwordHash": "TFJtUjKMN4dNcp5IuSwaeRkPwjwpkp9ZlqXuL/QHsQ3097QLHnZWccWt2yThLa0Q5rmbuOOXoqzoHBZM8x3GpQ==",
      "salt": "RHQ5jbaxNJ1lDA==",
      "createdAt": "1648020072141",
      "disabled": false,
      "providerUserInfo": []
    }
]}
----

== Importing Users

Next up, import the user data. Here are the steps we need to take.

1. Set Up FusionAuth
2. Add the scrypt Hashing Plugin
3. Get the Script
4. Install Needed Gems
5. Use the Script
6. Verify the Import
7. The Final Destination of Imported Users

=== Set Up FusionAuth

include::docs/v1/tech/migration-guide/_set-up-fusionauth.adoc[]

==== Create a Test Tenant

include::docs/v1/tech/migration-guide/_create-test-tenant.adoc[]

==== Create a Test Application

include::docs/v1/tech/migration-guide/_create-test-application.adoc[]

==== Add an API Key

include::docs/v1/tech/migration-guide/_create-api-key.adoc[]

=== Add the scrypt Hashing Plugin

Most Firebase projects use a https://github.com/firebase/scrypt[modified
version of the scrypt algorithm] for password hashes. FusionAuth does
not support scrypt directly, but does

> https://fusionauth.io/docs/v1/tech/plugins/custom-password-hashing[support

Reviewer: ```suggestion
link:/docs/v1/tech/plugins/custom-password-hashing[support
```

We should keep all links to FusionAuth relative so that we can build this locally and click around without being sent off to prod.

plugins for custom hashing].

To extend FusionAuth to support scrypt, follow the instructions for
https://fusionauth.io/docs/v1/tech/plugins/writing-a-plugin[cloning,

> building and installing the plugins example repository]. Before building

Reviewer: Looks like this needs to be updated for the fusionauth-contrib repository.

Author: updated in the main plugins guide, but the old repo is also referred to in a few other places which we didn't update yet.

Reviewer: You mean outside of the migration guide section? Don't worry about that.

the project, navigate to the
`src/main/java/com/mycompany/fusionauth/plugins/ExampleFirebaseScryptPasswordEncryptor.java`
file to add in a few scrypt parameters, which we'll get from Firebase.

In Firebase, navigate to the Authentication panel, and click on the three
dots near the *Add User* button. Then select *Password hash
parameters*. This will open a modal with the parameters we need. Make a
note of them.

image::migration-guide/{migration_source_dir}/firebase-hash-parameters.png[Copying the Firebase hash parameters,width=1200]

Now open the
`src/main/java/com/mycompany/fusionauth/plugins/ExampleFirebaseScryptPasswordEncryptor.java`
file. Copy the Firebase parameters from above into the variables under
the `Firebase Scrypt Parameters` comment.

> 

Reviewer: ```suggestion
```

image::migration-guide/{migration_source_dir}/plugin-firebase-parameters.png[Adding the Firebase scrypt parameters to the FusionAuth plugin,width=1000]

After you have copied over the parameters, build and install the plugin
project as detailed

> https://fusionauth.io/docs/v1/tech/plugins/writing-a-plugin[in the plugins guide]. You may need to change the test case parameters in the

Reviewer: ```suggestion
link:/docs/v1/tech/plugins/writing-a-plugin[in the plugins guide]. You may need to change the test case parameters in the
```

file
`src/test/java/com/mycompany/fusionauth/plugins/ExampleFirebaseScryptPasswordEncryptorTest.java`
to match a known password, salt and hash from your Firebase
installation.

After installing the plugin, and restarting FusionAuth, navigate to the
Test Tenant you created earlier. Click the *edit* icon. Under the
*Password* tab, find the section *Cryptographic hash settings*. Choose
*example-salted-firebase-scrypt* as the *Scheme*.

image::migration-guide/{migration_source_dir}/set-hash-scheme.png[Set hash scheme,width=1200]

> 

Reviewer: ```suggestion

If that plugin doesn't show up, please review the link:/docs/v1/tech/plugins/writing-a-plugin#troubleshooting[plugin troubleshooting steps].

```

[NOTE.note]
====
The social account linking functionality will only work with FusionAuth versions above or equal to 1.28. The `fusionauth_client` library must be >= 1.28 as well.
====

You may or may not want to use the `-m` switch, which takes the Firebase Id for users without a social login and uses the same value for the FusionAuth user Id. If you have external systems reliant on the Firebase user identifier, set this. Doing so ensures imported users have the same Id as they did in Firebase. Otherwise, you can omit this switch.

When you run the script, you’ll see output like:

[source,shell,title=Import script output]
----
$ ruby ./import.rb -f http://localhost:9011 -k '...' -u user-data.json
FusionAuth Importer : Firebase
 > User file: user-data.json
 > Call FusionAuth to import users
 > Import success
Duplicate users 0
Import complete. 2 users imported.
----

==== Enhancing the Script

You may also want to migrate additional data. Currently, the following
attributes are migrated:

* `user_id`
* `email`
* `email_verified`
* `username`
* `insertInstant`
* the password hash and supporting attributes, if available
* `registrations`, if supplied

> The migrated user will have the Firebase tenant Id and original user Id

Reviewer: ```suggestion
The migrated user will have the Firebase project Id and original user Id
```

It's a project Id, correct?
